File: next/kmp/core/src/commonMain/kotlin/org/dweb_browser/core/ipc/helper/EndpointLifecycle.kt
package org.dweb_browser.core.ipc.helper

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Polymorphic
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import org.dweb_browser.helper.OrderBy

/**
 * Ipc生命周期控制
 */
@Serializable
@SerialName(ENDPOINT_MESSAGE_TYPE_LIFECYCLE)
data class EndpointLifecycle(
  @Polymorphic val state: EndpointLifecycleState,
  override val order: Int = DEFAULT_ORDER,
) : EndpointRawMessage, EndpointMessage, OrderBy {
  companion object {
    const val DEFAULT_ORDER = -1
  }
}

@OptIn(ExperimentalSerializationApi::class)
@Serializable
@Polymorphic
@JsonClassDiscriminator("name")
sealed interface EndpointLifecycleState

const val ENDPOINT_LIFECYCLE_STATE_INIT = "init"
const val ENDPOINT_LIFECYCLE_STATE_OPENING = "opening"
const val ENDPOINT_LIFECYCLE_STATE_OPENED = "opened"
const val ENDPOINT_LIFECYCLE_STATE_CLOSING = "closing"
const val ENDPOINT_LIFECYCLE_STATE_CLOSED = "closed"

enum class EndpointProtocol {
  JSON, CBOR,
//  Protobuf,
}

@Serializable
@SerialName(ENDPOINT_LIFECYCLE_STATE_INIT)
data object EndpointLifecycleInit : EndpointLifecycleState

@Serializable
@SerialName(ENDPOINT_LIFECYCLE_STATE_OPENING)
data class EndpointLifecycleOpening(
  val subProtocols: Set<EndpointProtocol> = setOf(),
  // 双方各自贡献一个id，然后组合成一个大的id
  val sessionIds: List<String>,
) : EndpointLifecycleState

@Serializable
@SerialName(ENDPOINT_LIFECYCLE_STATE_OPENED)
data class EndpointLifecycleOpened(
  val subProtocols: Set<EndpointProtocol> = setOf(),
  val sessionPair: String,
) : EndpointLifecycleState

@Serializable
@SerialName(ENDPOINT_LIFECYCLE_STATE_CLOSING)
data class EndpointLifecycleClosing(val reason: String? = null) : EndpointLifecycleState

@Serializable
@SerialName(ENDPOINT_LIFECYCLE_STATE_CLOSED)
data class EndpointLifecycleClosed(val reason: String? = null) : EndpointLifecycleState

File: desktop-dev/src/core/ipc/endpoint/EndpointLifecycle.ts
import { ENDPOINT_LIFECYCLE_STATE, ENDPOINT_PROTOCOL } from "./internal/EndpointLifecycle";
import { ENDPOINT_MESSAGE_TYPE, endpointMessageBase } from "./internal/EndpointMessage";
export { ENDPOINT_LIFECYCLE_STATE, ENDPOINT_PROTOCOL } from "./internal/EndpointLifecycle";

export type $EndpointLifecycle<T extends $EndpointLifecycleState = $EndpointLifecycleState> = ReturnType<
  typeof _endpointLifecycle<T>
>;
const _endpointLifecycle = <T extends $EndpointLifecycleState>(state: T) =>
  ({
    ...endpointMessageBase(ENDPOINT_MESSAGE_TYPE.LIFECYCLE),
    state,
  } as const);

export const endpointLifecycle = Object.assign(_endpointLifecycle, {
  equals: (a: $EndpointLifecycle, b: $EndpointLifecycle) => {
    if (a.state.name !== b.state.name) {
      return false;
    }
    if (a.state.name === ENDPOINT_LIFECYCLE_STATE.CLOSING) {
      return a.state.reason === (b.state as $EndpointLifecycleClosing).reason;
    }
    if (a.state.name === ENDPOINT_LIFECYCLE_STATE.CLOSED) {
      return a.state.reason === (b.state as $EndpointLifecycleClosed).reason;
    }
    if (a.state.name === ENDPOINT_LIFECYCLE_STATE.OPENING || a.state.name === ENDPOINT_LIFECYCLE_STATE.OPENED) {
      return JSON.stringify(a.state) === JSON.stringify(b.state);
    }
    return true;
  },
});

export type $EndpointLifecycleState =
  | $EndpointLifecycleInit
  | $EndpointLifecycleOpening
  | $EndpointLifecycleOpened
  | $EndpointLifecycleClosing
  | $EndpointLifecycleClosed;
export type $EndpointLifecycleStateBase<S extends ENDPOINT_LIFECYCLE_STATE> = ReturnType<
  typeof endpointLifecycleStateBase<S>
>;
export const endpointLifecycleStateBase = <S extends ENDPOINT_LIFECYCLE_STATE>(name: S) => ({ name } as const);
export type $EndpointLifecycleInit = ReturnType<typeof endpointLifecycleInit>;
export type $EndpointLifecycleOpening = ReturnType<typeof endpointLifecycleOpening>;
export type $EndpointLifecycleOpened = ReturnType<typeof endpointLifecycleOpend>;
export type $EndpointLifecycleClosing = ReturnType<typeof endpointLifecycleClosing>;
export type $EndpointLifecycleClosed = ReturnType<typeof endpointLifecycleClosed>;

export const endpointLifecycleInit = () => endpointLifecycleStateBase(ENDPOINT_LIFECYCLE_STATE.INIT);
export const endpointLifecycleOpening = (subProtocols: Iterable<ENDPOINT_PROTOCOL>) =>
  ({
    ...endpointLifecycleStateBase(ENDPOINT_LIFECYCLE_STATE.OPENING),
    subProtocols: [...subProtocols],
  } as const);
export const endpointLifecycleOpend = (subProtocols: Iterable<ENDPOINT_PROTOCOL>) =>
  ({
    ...endpointLifecycleStateBase(ENDPOINT_LIFECYCLE_STATE.OPENED),
    subProtocols: [...subProtocols],
  } as const);
export const endpointLifecycleClosing = (reason?: string) =>
  ({
    ...endpointLifecycleStateBase(ENDPOINT_LIFECYCLE_STATE.CLOSING),
    reason,
  } as const);
export const endpointLifecycleClosed = (reason?: string) =>
  ({
    ...endpointLifecycleStateBase(ENDPOINT_LIFECYCLE_STATE.CLOSED),
    reason,
  } as const);
